### ISSUE TITLE
Memory Issues with Default Librispeech ASR task

### ISSUE THREAD

**Issue by Andy-Eum** [espnet/espnet#2240]
When trying to run stage 5 of the default Librispeech Transformer Model, the GCP instance that it was run on crashes.

I found out the issue is that it uses over 60 GBs of memory when tried with nj 8 - understandable, so tried it again with nj 1 - but the same issue happens.

The memory usage increases in a linear format along the percentage of decoding done when viewed via log and htop - for example, if 30/600 tasks were done at 5GB used out of 60GB, 60/600 tasks would be done with 10GB used out of 60GB - so it would crash near the 360/600 mark.

These are some things I tried:

1. Change nj to 8 to 1 - same issue, same propreties

2. Tried changing the gpu (was using ngpu 8)  via the --stage 5 --ngpu 0 option - same issue, same properties

3. Tried detaching the gpu (removed gpus via the GCP instance options) -     same issue, but not the same properties (doesn't scale linearly, yet the problem of consistent memory increase is still there)

4. Tried changing the cpu in the GCP n1 version to the memory-emphasized n2 version, along with additional RAM --  same issue, but inference did work - the maximum ram used was near 100 GB

 All batch configurations and such were kept to their default (decode_pytorch_transformer.yaml) , and the only change inside run.sh was the assignment of new paths along with # in front of rnnlm in stage 5 to not use the language model. The train.yaml has been slightly modified - but this shouldn't cause a problem in the decoding process of stage 5, right? 

Currently, the inference is being done with 200 GB of memory just in case, with no GPUs attached also. Is the default asr stage 5 task on the librispeech dataset a process that takes up more than 60 GB of memory?

Any insights into this problem would be appreciated. Thanks in advance.

**Comment by sw005320** [comment]
I just checked the recent Librispeech Transformer Model, but I did not observe such huge memory consumption. Could you paste the decoding log file header?

**Comment by Andy-Eum** [comment]
This is from decode.1.log with nj 4.

``# asr_recog.py --config conf/decode.yaml --ngpu 0 --backend pytorch --batchsize 0 --recog-json /audio/data/librispeech/data/dump/dev_clean/deltafalse/split4utt/data_unigram5000.1.json --result-label /audio/data/librispeech/exp/train_960_pytorch_mytrain/decode_dev_clean_model.acc.best_decode_lm/data.1.json --model /audio/data/librispeech/exp/train_960_pytorch_mytrain/results/model.acc.best --rnnlm exp/train_rnnlm_pytorch_lm_unigram5000_ngpu8//audio/data/rnnlm.model.best --api v2 
# Started at Fri Jul 31 02:49:38 UTC 2020
#
2020-07-31 02:49:39,161 (asr_recog:253) INFO: python path = (None)
2020-07-31 02:49:39,161 (asr_recog:258) INFO: set random seed = 1
2020-07-31 02:49:39,161 (asr_recog:269) INFO: backend = pytorch
2020-07-31 02:49:40,841 (recog:34) WARNING: experimental API for custom LMs is selected by --api v2
2020-07-31 02:49:40,841 (deterministic_utils:26) INFO: torch type check is disabled
2020-07-31 02:49:40,842 (asr_utils:478) INFO: reading a config file from /audio/data/librispeech/exp/train_960_pytorch_mytrain/results/model.json
2020-07-31 02:49:40,843 (asr_init:169) WARNING: reading model parameters from /audio/data/librispeech/exp/train_960_pytorch_mytrain/results/model.acc.best
2020-07-31 02:49:41,096 (encoder:132) INFO: encoder self-attention layer type = self-attention
2020-07-31 02:49:41,483 (decoder:112) INFO: decoder self-attention layer type = self-attention
2020-07-31 02:49:41,573 (ctc:35) WARNING: CTC was set to builtin due to PyTorch version.
2020-07-31 02:49:42,166 (asr_utils:478) INFO: reading a config file from exp/train_rnnlm_pytorch_lm_unigram5000_ngpu8//audio/data/model.json
2020-07-31 02:49:42,175 (encoder:132) INFO: encoder self-attention layer type = self-attention
2020-07-31 02:49:42,852 (recog:123) INFO: Decoding device=cpu, dtype=torch.float32``

Memory Usage with htop is shown as follows:

![image](https://user-images.githubusercontent.com/57177551/88995171-fae7cc80-d324-11ea-9762-c3846c77f5aa.png)
this was at the 150/676 mark.

![image](https://user-images.githubusercontent.com/57177551/88995226-1521aa80-d325-11ea-964a-c018313a5785.png)
this was at the 180/676 mark.

Although the logs and pictures were taken at nj 4, the same happens with nj 1, along with the memory peaking near 120GB as it finishes.

**Comment by sw005320** [comment]
Hmm, I tried to use the same setup with you and I still did not observe such an issue (around 2GB for one job). it might be due to your environment.
Could you use https://github.com/espnet/espnet/blob/master/utils/show_result.sh and show us your environments like https://github.com/espnet/espnet/blob/master/egs/aidatatang_200zh/asr1/RESULTS.md#environments ?

**Comment by Andy-Eum** [comment]
```
# RESULTS
## Environments
- date: `Mon Aug  3 01:09:29 UTC 2020`
- python version: `3.7.3 (default, Mar 27 2019, 22:11:17)  [GCC 7.3.0]`
- espnet version: `espnet 0.5.3`
- chainer version: `chainer 6.0.0`
- pytorch version: `pytorch 1.4.0`
- Git hash: `53468ac2e64146bf2a3a03eb77f44a4a32f72c59`
  - Commit date: `Wed Jan 29 06:12:45 2020 +0000`
```
Here is the environment settings ran with show_result.sh. 
2GB is nowhere near the memory I'm using currently to run stage 5.

**Comment by sw005320** [comment]
Could you switch to use the latest espnet?

**Comment by Andy-Eum** [comment]
Was there any problems with memory in any of the previous versions? Especially around stage 5? 
A project with other modified programs to work with this version is going on right now, and hence would rather not use the latest espnet. If the problem is due to a lower version, I'll eventually try to change all the settings.

**Comment by sw005320** [comment]
We’ve never had such reports, and I also did not have that issue personally. At that time I used the old pytorch (1.01), and many users were like that at that time.
However, some combinations (espnet and pytorch versions, but there are other possible combinations come from the OS, cuda, and cudnn versions) may cause such issues.
Due to our limited resource, we cannot test all espnet and pytorch versions. I hope you understand our situations.

Another possibility is to downgrade your pytorch versions instead of updating espnet to the latest versions.
For example, I found the following issue. This might be related to your problem.
https://github.com/espnet/espnet/issues/1740
